### PR TITLE
chore: implements YAML diff writer for `map` and scalar `seq`

### PR DESCRIPTION
**Pull request by Lou1415926**
That writer works for `map` node, `scalar` node and  `seq` nodes whose content are not `map`.

The writer is on parity with the differ in `diff.go`. The to do for both are the same:
1. Handle "(x unchanged items)"
2. Handle `seq` with `map` content

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/diff/write.go
Comment (Lou1415926) [inline internal/pkg/template/diff/write.go:1]: Although this PR needs to be rebased after #4632 is merged, and has to be merged after that one - the logics in `formatter.go` and `write.go` should be reviewable already!

Edit: rebased1!!!

File: internal/pkg/template/diff/diff.go
Comment (dannyrandall) [inline internal/pkg/template/diff/diff.go]: just kinda laughing at these names as i was reading through the PR... too bad we can't use the function names (`key`/`children`)

File: internal/pkg/template/diff/formatter.go
Comment (dannyrandall) [inline internal/pkg/template/diff/formatter.go:1]: i'm assuming tests will come for this file so i won't hold up this PR for that - just will help in seeing what the formats look like :)

File: internal/pkg/template/diff/write.go
Comment (dannyrandall) [inline internal/pkg/template/diff/write.go]: any chance we could use `color` to highlight delete's red and insert's green (or something?)

**Review by dannyrandall — APPROVED** [review]

File: internal/pkg/template/diff/formatter.go
Comment (Lou1415926) [reply internal/pkg/template/diff/formatter.go:1]: do you think we need one 👀  i was hoping `write_test.go` is enough to cover everything that we care about 😂

File: internal/pkg/template/diff/formatter.go
Comment (efekarakus) [inline internal/pkg/template/diff/formatter.go]: 😂 
```suggestion
func (f *seqItemFormatter) formatYAML(node *yaml.Node) ([]byte, error) {
```

File: internal/pkg/template/diff/formatter.go
Comment (efekarakus) [inline internal/pkg/template/diff/formatter.go]: nit: This makes me think that maybe we should rename the methods to this? 💭 
```suggestion
	return fmt.Sprintf("%s: %s -> %s", node.key(), node.oldYAML().Value, node.newYAML().Value)
```

File: internal/pkg/template/diff/formatter.go
Comment (efekarakus) [inline internal/pkg/template/diff/formatter.go]: nit:
```suggestion
func prefixByFn(prefix string) func(line string) string {
```

File: internal/pkg/template/diff/formatter.go
Comment (efekarakus) [inline internal/pkg/template/diff/formatter.go]:
```suggestion
func indentByFn(count int) func(line string) string {
```

File: internal/pkg/template/diff/write.go
Comment (efekarakus) [inline internal/pkg/template/diff/write.go]: Out of curiosity, was there a reason why you didn't prefer `Write` to be a method on `Tree`?
e.g.
```go
// Write writes a visual representation of the diff tree to w.
func (t *Tree) Write(w io.Writer) error
```

**Review by efekarakus — COMMENTED** [review]
I haven't finished reviewing but I'm publishing what I have so far!

File: internal/pkg/template/diff/formatter.go
Comment (dannyrandall) [reply internal/pkg/template/diff/formatter.go:1]: oh good point lol. prob fine without!

File: internal/pkg/template/diff/write.go
Comment (Lou1415926) [reply internal/pkg/template/diff/write.go]: I thought of `Tree` as the data storage structure, and how to render the data is a different responsibility. However, it could on the other hand be argued that rendering is also a part of the spec that the tree has to provide. I feel like there isn't a clear winner which is better, and I don't have a super concrete reason for going the way I chose. The real thought process was that when I realized I'd need to pass `w io.Writer` around in all these private functions, I thought maybe it's a sign that it should have been a struct on its own 😂

Comment (efekarakus) [reply internal/pkg/template/diff/write.go]: oh it's fine, I was just trying to think if we can reduce the API surface of the package. We can still have a struct like this:
```

func (t *Tree) Write(w io.Writer) error {
  tw := &treeWriter{t, w}
  return tw.Write()
}

type treeWriter struct {
   tree *Tree
   out   io.Writer
}
```

File: internal/pkg/template/diff/write.go
Comment (efekarakus) [inline internal/pkg/template/diff/write.go]: I wonder if we should instead just no write anything? And let the client decide how to display info back.

For example, here is what I'm imagining happens in the `cli` pkg:
```go
sb := new(strings.Builder)
tree, err := diff.Parse(deployedTemplate, pkgTemplate)
if err != nil { return err }
if err := tree.Write(sb); err != nil { return err }

out := sb.String()
if len(out) == "" {
   log.Info("There are no changes between the deployed stack and the local manifest for %q", o.name)
}
fmt.Println(out)
```

File: internal/pkg/template/diff/write.go
Comment (efekarakus) [inline internal/pkg/template/diff/write.go:82]: this is so nice 🤩

File: internal/pkg/template/diff/write_test.go
Comment (efekarakus) [inline internal/pkg/template/diff/write_test.go]: nit: would you mind starting the multiline strings with a new line, to make it easier to see the end outcome e.g.
```suggestion
			curr: `
Mary:
```

File: internal/pkg/template/diff/write_test.go
Comment (efekarakus) [inline internal/pkg/template/diff/write_test.go:75]: Isn't this testcase captured by "list with a scalar value changed" below?

File: internal/pkg/template/diff/write_test.go
Comment (efekarakus) [inline internal/pkg/template/diff/write_test.go:92]: we'll have the "# of unmodified items" later right?

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/template/diff/write_test.go
Comment (Lou1415926) [reply internal/pkg/template/diff/write_test.go:75]: yeahhh I left it here just in case I have the time to implement the logic that writes ⬇️  😂 I think of this as a nice to have though 😂 
```
bear(moved from 2 to 3), dog (moved from 3 to 2)
```

(not important but instead of "scalar value changed", it'd actually be an insertion (at the new position)  + a deletion (from the old position))

File: internal/pkg/template/diff/write_test.go
Comment (Lou1415926) [reply internal/pkg/template/diff/write_test.go:92]: yessssss

**Pull request merged**
